### DISCUSSION TITLE
Using the same HMAC state without re-init

### DISCUSSION THREAD

**Discussion by Tachi107** [jedisct1/libsodium#1240] [Q&A] [answered]
Hi, first of all thank you for this awesome library!

I'm currently looking into using the HMAC-SHA-2 multi-part API, and one limitation I noticed is that it doesn't seem to be possible to reuse the same HMAC state without calling `crypto_auth_hmacsha256_init()` again. I'm looking for a function similar to MbedTLS' [`mbedtls_md_hmac_reset()`](https://web.archive.org/web/20210623172227/https://tls.mbed.org/api/md_8h.html#a3a5c64b5bda0f294e917ccd5a8ca234e), which can be used after calling `_final()` to perform another HMAC operation with the same key.

Would it make sense to have such an API in libsodium? Should I just call `_init()` again with the same key? Would reusing the same state provide any significant performance/efficiency advantage?

Thanks again :)

PS: I tried to subscribe to the mailing list mentioned in the README, but nothing happened. Is it still in use?

**Comment by jedisct1** [comment, marked as answer]
Hi!

The `crypto_auth_*_state` values can be copied. Libsodium doesn't do dynamic allocations, so there are no pointers inside and you can just copy them as a blob of bytes.

So, before calling `_final()`, copy the state, and reuse it afterwards as needed.

**Reply by jedisct1** [reply]
Oh, I misread your question and thought you wanted to finalize the tag before the entire message was absorbed.

But the answer is the same :) The state can be copied after `_init()` if you want to reuse it.

That being said `_init()` is a very fast operation, so there is no significant advantage in doing so.

**Comment by Tachi107** [comment]
Il giorno mar 27 dic 2022 alle 04:20:53 -08:00:00, Frank Denis 
***@***.***> ha scritto:
> The crypto_auth_*_state values can be copied. Libsodium doesn't do 
> dynamic allocations, so there are no pointers inside and you can just 
> copy them as a blog of bytes.
> 
> So, before calling _final(), copy the state, and reuse it afterwards 
> as needed.
> 

Hi Frank, thanks for the fast reply! I'll do as you suggested :)
